### PR TITLE
[RHELC-1228, RHELC-1231] Improve detection of RHSM credentials in subscription check

### PR DESCRIPTION
**Pull request by pr-watson**
This PR adds a check to see if SCA is not enabled and if there is no subscription attached. In the event these conditions are met we attempt to auto attach a subscription for the customer. If the auto attachment fails we have an ERROR result `NO_ACCESS_TO_RHEL_REPOS`  to let the user know that the system is registered with an RHSM account that has SCA disabled but no subscription is attached and the auto attachment was unsuccessful. New functions is_subscription_attached() and auto_attach_subscription() have been made to facilitate these checks. Also made a new function _is_sca_enabled so we can detect if sca is enabled in both subscription.py and pre_ponr_changes/subscription.py. 

Jira Issues: 
* [RHELC-1228](https://issues.redhat.com/browse/RHELC-1228)
* [RHELC-1231](https://issues.redhat.com/browse/RHELC-1231)

Checklist

- [x] PR has been tested manually in a VM (either author or reviewer)
- [x] Jira issue has been made public if possible
- [x] `[RHELC-]` is part of the PR title 
- [x] GitHub label has been added to help with Release notes 
- [x] PR title explains the change from the user's point of view
- [x] Code and tests are documented properly
- [x] The commits are squashed to as few commits as possible (without losing data) 
- [ ] When merged: Jira issue has been updated to `Release Pending` if relevant

### CODE REVIEW

File: convert2rhel/subscription.py [hunk 1]
```diff
@@ -560,6 +560,20 @@ def install_rhel_subscription_manager(pkgs_to_install, pkgs_to_upgrade=None):
...
+def is_sca_enabled():
+    """
+    Check if Simple Content Access has been enabled for this system.
+    :returns: True if Simple Content Access is enabled.
+    :rtype: bool
+    """
+    # check if SCA is enabled
+    output, _ = utils.run_subprocess(["subscription-manager", "status"], print_output=False)
+    if "content access mode is set to simple content access." in output.lower():
+        loggerinst.info("Simple Content Access is enabled, skipping subscription attachment")
```
Comment (SpyTec) [inline convert2rhel/subscription.py:572]: I would not put in the `skipping subscription attachment` here in case we ever re-use this function somewhere else, it might give a confusing message, I would put the log back where it was if anything

File: convert2rhel/actions/pre_ponr_changes/subscription.py [hunk 2]
```diff
@@ -195,7 +195,15 @@ def run(self):
...
                 raise
 
-            logger.warning("No rhsm credentials given to subscribe the system. Skipping the subscription step.")
+            if subscription._is_registered() and not subscription.is_sca_enabled():
+                self.set_result(
+                    level="ERROR",
+                    id="SYSTEM_REGISTERED_WITHOUT_SCA",
+                    title="Registered with RHSM but without SCA enabled",
+                    description="This system has been registered with Red Hat Subscription Manager but Simple Content Access is not enabled.",
+                    remediation="To resolve this error please enable Simple Content Access and run the conversion again.",
```
Comment (SpyTec) [inline convert2rhel/actions/pre_ponr_changes/subscription.py:204]: How is the recommendation when it comes to writing SCA vs Simple Content Access? Can we use SCA in titles or should we use the full name? I'll ping Miriam

**Review by SpyTec — CHANGES_REQUESTED** [review]

File: convert2rhel/actions/pre_ponr_changes/subscription.py (hunk 2 above)
Comment (SpyTec) [reply convert2rhel/actions/pre_ponr_changes/subscription.py:204]: Both Pino and Miriam says it is spelled out first, then abbreviated, like `Simple Content Access (SCA)`. So I would change to something like this

```suggestion
                    id="SYSTEM_REGISTERED_WITHOUT_SCA",
                    title="Registered with RHSM but without Simple Content Access (SCA) enabled",
                    description="This system has been registered with Red Hat Subscription Manager but Simple Content Access is not enabled.",
                    remediation="To resolve this error please enable Simple Content Access and run the conversion again.",
```

**Comment by SpyTec** [comment]
Unit tests are failing, and we need to update `remediation` to `remediations`

File: convert2rhel/actions/pre_ponr_changes/subscription.py [hunk 3]
```diff
@@ -195,7 +195,15 @@ def run(self):
...
                 raise
 
-            logger.warning("No rhsm credentials given to subscribe the system. Skipping the subscription step.")
+            if subscription._is_registered() and not subscription.is_sca_enabled():
+                self.set_result(
+                    level="ERROR",
+                    id="SYSTEM_REGISTERED_WITHOUT_SCA",
+                    title="Registered with RHSM but without Simple Content Access (SCA) enabled",
+                    description="This system has been registered with Red Hat Subscription Manager but Simple Content Access is not enabled.",
+                    remediation="To resolve this error please enable Simple Content Access and run the conversion again.",
```
Comment (r0x0d) [inline convert2rhel/actions/pre_ponr_changes/subscription.py:204]: I think it would be good to include some link to in the remediation(?) that show the user where they could neable the SCA. 

Also, don't forget to update the keyword here.
```suggestion
                    remediations="To resolve this error please enable Simple Content Access and run the conversion again.",
```

File: convert2rhel/subscription.py [hunk 4]
```diff
@@ -560,6 +560,19 @@ def install_rhel_subscription_manager(pkgs_to_install, pkgs_to_upgrade=None):
     backup.backup_control.push(installed_pkg_set)
 
 
+def is_sca_enabled():
+    """
+    Check if Simple Content Access has been enabled for this system.
+    :returns: True if Simple Content Access is enabled.
+    :rtype: bool
```
Comment (r0x0d) [inline convert2rhel/subscription.py:567]:
```suggestion
    Check if Simple Content Access has been enabled for this system.
    
    :returns: True if Simple Content Access is enabled.
    :rtype: bool
```

**Review by r0x0d — COMMENTED** [review]
Looks good to me. Only one comment.

File: convert2rhel/actions/pre_ponr_changes/subscription.py (hunk 3 above)
Comment (pr-watson) [reply convert2rhel/actions/pre_ponr_changes/subscription.py:204]: Good point i'll add this

**Review by r0x0d — COMMENTED** [review]
Looks good!

File: convert2rhel/subscription.py (hunk 4 above)
Comment (SpyTec) [reply convert2rhel/subscription.py:567]: Docstring syntax requires a space here

File: convert2rhel/subscription.py [hunk 5]
```diff
@@ -569,7 +582,7 @@ def attach_subscription():
 
     # check if SCA is enabled
     output, _ = utils.run_subprocess(["subscription-manager", "status"], print_output=False)
-    if "content access mode is set to simple content access." in output.lower():
+    if is_sca_enabled():
```
Comment (SpyTec) [inline convert2rhel/subscription.py:585]: The status call here can be removed right?

File: convert2rhel/actions/pre_ponr_changes/subscription.py [hunk 6]
```diff
@@ -195,7 +195,15 @@ def run(self):
                     return
                 raise
 
-            logger.warning("No rhsm credentials given to subscribe the system. Did not perform the subscription step.")
+            if subscription._is_registered() and not subscription.is_sca_enabled():
```
Comment (SpyTec) [inline convert2rhel/actions/pre_ponr_changes/subscription.py:198]: `_is_registered` is used outside the module, please change it to be `is_registered` to not indicate it's a private function, as it isn't anymore

**Review by SpyTec — CHANGES_REQUESTED** [review]

File: convert2rhel/subscription.py (hunk 5 above)
Comment (pr-watson) [reply convert2rhel/subscription.py:585]: Good catch I will remove it

File: convert2rhel/actions/pre_ponr_changes/subscription.py (hunk 6 above)
Comment (r0x0d) [reply convert2rhel/actions/pre_ponr_changes/subscription.py:198]: +1

**Comment by SpyTec** [comment]
Unit tests failing

**Comment by SpyTec** [comment]
I feel like the integration tests here are failing due to SCA but not sure. @pr-watson?

**Comment by pr-watson** [comment]
> I feel like the integration tests here are failing due to SCA but not sure. @pr-watson?
> 
> ```
> 
> [2024-02-20T15:23:33+0000] TASK - [Convert: Subscription Manager - Reload configuration] *************
> 
> [2024-02-20T15:23:33+0000] DEBUG - Calling command 'subscription-manager refresh'
> 
> subscription-manager has reloaded its configuration.
> 
> [2024-02-20T15:23:39+0000] DEBUG - Checking whether the host was registered.
> 
> [2024-02-20T15:23:39+0000] DEBUG - Calling command 'subscription-manager identity'
> 
> system identity: [removed]
> 
> name: [removed].testing-farm
> 
> org name: [removed]
> 
> org ID: [removed]
> 
> [2024-02-20T15:23:45+0000] DEBUG - Host was registered.
> 
> [2024-02-20T15:23:45+0000] DEBUG - Calling command 'subscription-manager status'
> 
> ERROR - (ERROR) SUBSCRIBE_SYSTEM::SYSTEM_REGISTERED_WITHOUT_SCA - Registered with RHSM but without SCA enabled
> 
>  Description: This system has been registered with Red Hat Subscription Manager but Simple Content Access is not enabled.
> 
>  Diagnosis: N/A
> 
>  Remediations: To resolve this error please enable Simple Content Access at https://access.redhat.com/management/ and run the conversion again.
> 
> 
> 
> ERROR - Skipped ENSURE_KERNEL_MODULES_COMPATIBILITY. Skipped because SUBSCRIBE_SYSTEM was not successful
> 
> ERROR - Skipped VALIDATE_PACKAGE_MANAGER_TRANSACTION. Skipped because ENSURE_KERNEL_MODULES_COMPATIBILITY and SUBSCRIBE_SYSTEM were not successful
> 
> CRITICAL - Conversion failed.
> 
> Writing breadcrumbs to '/etc/migration-results'.
> 
> Writing RHSM custom facts to '/etc/rhsm/facts/convert2rhel.facts'.
> 
> [2024-02-20T15:23:51+0000] DEBUG - Traceback (most recent call last):
> 
>   File "/usr/lib/python3.6/site-packages/convert2rhel/main.py", line 140, in main_locked
> 
>     loggerinst.critical("Conversion failed.")
> 
>   File "/usr/lib/python3.6/site-packages/convert2rhel/logger.py", line 250, in _critical
> 
>     sys.exit(msg)
> 
> SystemExit: Conversion failed.
> 
> 
> 
> WARNING - Abnormal exit! Performing rollback ...
> ```

Yes this is a new Error result added so the integration tests need to be updated to accommodate it

**Comment by pr-watson** [comment]
I've added an update to test_pre_registered.py to expect the new error result when running c2r on a preregistered system without SCA enabled. Some other integration tests likely need to be updated

File: tests/integration/tier0/destructive/conversion-method/test_pre_registered.py [hunk 7]
```diff
@@ -4,5 +4,5 @@
 @pytest.mark.test_pre_registered_conversion
 def test_run_conversion_pre_registered(convert2rhel, pre_registered):
     with convert2rhel("-y --debug") as c2r:
-        c2r.expect("Conversion successful!")
-    assert c2r.exitstatus == 0
+        c2r.expect("SUBSCRIBE_SYSTEM::SYSTEM_REGISTERED_WITHOUT_SCA")
+    assert c2r.exitstatus == 1
```
Comment (danmyway) [inline tests/integration/tier0/destructive/conversion-method/test_pre_registered.py:8]: This doesn't make sense at this point.
The test is supposed to validate, that it's possible to convert a pre-registered system.
This seems to be just a fix for the sake of the test passing.
Please revert this change, the fail would be expected since the behavior changed and we will address that separately.
Thanks

**Review by danmyway — DISMISSED** [review]

File: convert2rhel/subscription.py [hunk 8]
```diff
@@ -561,6 +561,20 @@ def install_rhel_subscription_manager(pkgs_to_install, pkgs_to_upgrade=None):
...
 
 
+def is_sca_enabled():
+    """
+    Check if Simple Content Access has been enabled for this system.
+
+    :returns: True if Simple Content Access is enabled.
+    :rtype: bool
+    """
+    # check if SCA is enabled
```
Comment (r0x0d) [inline convert2rhel/subscription.py:571]: Maybe this comment is not adding any value to the function, but not a blocker.

File: convert2rhel/unit_tests/actions/pre_ponr_changes/subscription_test.py [hunk 9]
```diff
@@ -257,6 +267,40 @@ def test_subscribe_system_no_rhsm_option_detected(
...
         assert expected.issubset(subscribe_system_instance.messages)
 
+    def test_subscribe_system_not_registered(self, global_tool_opts, subscribe_system_instance, monkeypatch):
+
+        monkeypatch.setattr(subscription, "should_subscribe", partial(toolopts._should_subscribe, global_tool_opts))
+        monkeypatch.setattr(utils, "run_subprocess", RunSubprocessMocked(return_code=1))
+
+        with pytest.raises(RefreshSubscriptionManagerError):
+            subscribe_system_instance.run()
+            print(subscribe_system_instance.result)
```
Comment (r0x0d) [inline convert2rhel/unit_tests/actions/pre_ponr_changes/subscription_test.py:277]:
```suggestion
```

File: tests/integration/tier0/destructive/conversion-method/test_pre_registered.py (hunk 7 above)
Comment (pr-watson) [reply tests/integration/tier0/destructive/conversion-method/test_pre_registered.py:8]: I'll revert this change thanks

**Review by bocekm — CHANGES_REQUESTED** [review]
The code does not solve one of the acceptance criteria:
"An errors is raised when we detect that the system is not registered with sub-man AND no RHSM credentials were passed to convert2rhel"

**Comment by pr-watson** [comment]
> The code does not solve one of the acceptance criteria: "An errors is raised when we detect that the system is not registered with sub-man AND no RHSM credentials were passed to convert2rhel"

@bocekm about 2 months ago I discussed with you in slack about why this wasn't implemented. I did implement an error for the acceptance criteria but in testing I found that the first error caught in the try/except - SYSTEM_NOT_REGISTERED handles the case where the system is not registered with sub-man AND no RHSM credentials were passed to convert2rhel so it was deemed unnecessary to add since it's already handled

**Comment by pr-watson** [comment]
I've added the new functions and errors requested in the updates to the ticket. Still need to add unit tests and test in a VM

File: convert2rhel/subscription.py [hunk 10]
```diff
@@ -565,6 +569,43 @@ def install_rhel_subscription_manager(pkgs_to_install, pkgs_to_upgrade=None):
...
+    if "no consumed subscription pools were found." in output.lower():
+        return False
+    return True
+
+
+def auto_attach_subscription():
+    """
+    Execute 'subscription-manager attach --auto' to auto attach a subscription. If it fails raise
+    SubscriptionAutoAttachmentError.
+    """
```
Comment (r0x0d) [inline convert2rhel/subscription.py:603]:
```suggestion
    """
    Execute 'subscription-manager attach --auto' to auto attach a subscription. 
    
    :raises: SubscriptionAutoAttachmentError: If sub-man command fails to attach a subscription automatically.
    """
```

**Review by danmyway — CHANGES_REQUESTED** [review]
To be honest, I don't really follow what is supposed to be achieved here.
Do we disallow the non-SCA account?
If so, why is the non-SCA account/manually attaching the pool even considered in the remediation?
If not, why do we modify the test case covering this exact scenario? 
* non-SCA account registered, pool attached manually prior to the conversion
* I have tested the scenario, and it works fine. Where does the impulse to change the test behavior come from?

File: tests/integration/tier0/non-destructive/subscription-manager/main.fmf [hunk 11]
```diff
@@ -90,3 +90,31 @@ tag+:
             pytest -svv -m test_unregistered_no_credentials
         tag+:
             - unregistered-no-credentials
+    /no_sca_not_subscribed:
+        summary+: |
+            The subscription is auto-attached on a pre-registered system
```
Comment (kokesak) [inline tests/integration/tier0/non-destructive/subscription-manager/main.fmf:95]:
```suggestion
            The subscription is auto-attached by the convert2rhel utility on a pre-registered system
```

File: tests/integration/tier0/non-destructive/subscription-manager/test_sub_man_pre_register.py [hunk 12]
```diff
@@ -75,3 +75,43 @@ def test_unregistered_no_credentials(shell, convert2rhel):
...
+    """
+    This test verifies that running conversion on pre-registered system
+    without an attached subscription will try auto attaching the subscription.
+    """
+    with convert2rhel("--debug") as c2r:
+        # We need to get past the data collection acknowledgement.
+        c2r.expect("Continue with the system conversion?")
+        c2r.sendline("y")
+
+        c2r.expect("We'll try to auto-attach a subscription")
```
Comment (kokesak) [inline tests/integration/tier0/non-destructive/subscription-manager/test_sub_man_pre_register.py:91]: Is there some additional message saying something that the auto-attachment was successful?

File: tests/integration/tier0/non-destructive/subscription-manager/test_sub_man_pre_register.py [hunk 13]
```diff
@@ -75,3 +75,43 @@ def test_unregistered_no_credentials(shell, convert2rhel):
...
+        c2r.expect("Continue with the system conversion?")
+        c2r.sendline("n")
+
```
Comment (kokesak) [inline tests/integration/tier0/non-destructive/subscription-manager/test_sub_man_pre_register.py:94]: Maybe we should add the verification that the pool is not attached after the rollback - your question on slack https://redhat-internal.slack.com/archives/C05EWQE33TK/p1712740634169379

File: tests/integration/tier0/non-destructive/subscription-manager/main.fmf (hunk 11 above)
Comment (danmyway) [reply tests/integration/tier0/non-destructive/subscription-manager/main.fmf:95]: I will try to reword this, but I'd still like to keep this around 50 characters :D

File: convert2rhel/actions/pre_ponr_changes/subscription.py [hunk 14]
```diff
@@ -185,7 +185,22 @@ def run(self):
...
+                try:
+                    subscription.auto_attach_subscription()
+                except subscription.SubscriptionAutoAttachmentError:
+                    self.set_result(
+                        level="ERROR",
+                        id="NO_ACCESS_TO_RHEL_REPOS",
+                        title="No access to RHEL repositories",
+                        description="The system can access RHEL repositories only with either Simple Content Access (SCA) enabled or with an attached subscription.",
+                        diagnosis="The system is registered with an RHSM account that has SCA disabled but no subscription is attached. Auto-attaching a subscription was not successful.",
+                        remediations="Either attach a subscription manually by running 'subscription-manager attach --pool <pool id>' prior to the conversion or enable Simple Content Access on your RHSM account (https://access.redhat.com/articles/simple-content-access).",
```
Comment (r0x0d) [inline convert2rhel/actions/pre_ponr_changes/subscription.py:201]: We also have an option in convert2rhel to set the pool id for the conversion, wouldn't the case here to tell the user to use our `--pool` instead of telling them to run subscription-manager? I was not following the work here closely, but, if we will unregister the system at rollback, then the system won't be registered in the first place, right?

File: convert2rhel/subscription.py [hunk 15]
```diff
@@ -535,7 +539,7 @@ def _set_connection_opts_in_config(self):
             loggerinst.info("Successfully set RHSM connection configuration.")
 
 
-def _is_registered():
+def is_registered():
```
[the remark this answers is not in the document]
Comment (pr-watson) [reply convert2rhel/subscription.py:542]: Good point let me change this

File: convert2rhel/actions/pre_ponr_changes/subscription.py (hunk 14 above)
Comment (pr-watson) [reply convert2rhel/actions/pre_ponr_changes/subscription.py:201]: This particular case is to error when the user registers their system beforehand using subscription-manager but with an account without SCA enabled and no subs attached. About them using `--pool`, both options are viable - @bocekm recommended using the subscription-manager route

Comment (r0x0d) [reply convert2rhel/actions/pre_ponr_changes/subscription.py:201]: Sounds good to me.

File: tests/integration/tier0/non-destructive/subscription-manager/test_sub_man_pre_register.py (hunk 12 above)
Comment (danmyway) [reply tests/integration/tier0/non-destructive/subscription-manager/test_sub_man_pre_register.py:91]: Done @kokesak

Comment (kokesak) [reply tests/integration/tier0/non-destructive/subscription-manager/test_sub_man_pre_register.py:91]: Danke!

**Review by kokesak — APPROVED** [review]
LGTM! Waiting for final change to remove the attached subscription during rollback

**Comment by SpyTec** [comment]
> LGTM (testing wise)! Waiting for final change to remove the attached subscription during rollback

@kokesak feel free to squash merge when you think the final change is done

**Review by danmyway — APPROVED** [review]
Looks good from my POV
